event_emitter/src/event.rs

```diff
@@ -0,0 +1,344 @@
+        Self::Bool(v)
+    }
+}
+
+impl From<&'static str> for FieldValue {
+    fn from(v: &'static str) -> Self {
+        Self::String(v.into())
+    }
+}
+
+impl From<String> for FieldValue {
+    fn from(v: String) -> Self {
+        Self::String(v.into())
+    }
+}
+
+/// Typed InfluxDB-style event.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Event<M>
+where
+    M: Into<&'static str>,
+{
+    measurement: M,
+    time: Time,
+    tags: BTreeMap<&'static str, Cow<'static, str>>,
+    fields: BTreeMap<&'static str, FieldValue>,
+}
+
+impl<M> Event<M>
+where
+    M: Into<&'static str>,
+{
+    /// Create new measurement.
+    pub fn new(measurement: M, time: Time) -> Self {
+        Self {
+            measurement,
+            time,
+            tags: BTreeMap::default(),
+            fields: BTreeMap::default(),
+        }
+    }
+
+    /// Measurement.
+    pub fn measurement(&self) -> &M {
+        &self.measurement
+    }
+
+    /// Timestamp.
+    pub fn time(&self) -> Time {
+        self.time
+    }
+
+    /// Get all tags.
+    ///
+    /// The iterator elements are sorted by tag name.
+    pub fn tags(&self) -> impl Iterator<Item = (&'static str, &str)> {
+        self.tags.iter().map(|(k, v)| (*k, v.as_ref()))
+    }
+
+    /// Get all fields.
+    ///
+    /// The iterator elements are sorted by field name.
+    pub fn fields(&self) -> impl Iterator<Item = (&'static str, &FieldValue)> {
+        self.fields.iter().map(|(k, v)| (*k, v))
+    }
+
+    /// Add new tag.
+    ///
+    /// # Panic
+    /// Panics if a tag or a field with the same name already exists. Also panics if a tag called `"time"` is used.
+    pub fn add_tag_mut<V>(&mut self, name: &'static str, value: V) -> &mut Self
+    where
+        V: Into<Cow<'static, str>>,
+    {
+        if name == "time" {
+            panic!("Cannot use a tag called 'time'.");
+        }
+
+        if self.fields.contains_key(name) {
+            panic!(
+                "Cannot use tag named '{name}' because a field with the same name already exists"
+            );
+        }
+
+        match self.tags.entry(name) {
+            Entry::Vacant(v) => {
+                v.insert(value.into());
+            }
+            Entry::Occupied(_) => {
+                panic!("Tag '{name}' already used.")
+            }
+        }
+
+        self
+    }
+
+    /// Add new tag.
+    ///
+    /// # Panic
+    /// Panics if a tag or a field with the same name already exists. Also panics if a tag called `"time"` is used.
+    pub fn add_tag_move<V>(mut self, name: &'static str, value: V) -> Self
+    where
+        V: Into<Cow<'static, str>>,
+    {
+        self.add_tag_mut(name, value);
+        self
+    }
+
+    /// Add a new field.
+    ///
+    /// # Panic
+    /// Panics if a tag or a field with the same name already exists. Also panics if a field called `"time"` is used.
+    pub fn add_field_mut<V>(&mut self, name: &'static str, value: V) -> &mut Self
```

Comment:
```suggestion
    pub fn add_field<V>(&mut self, name: &'static str, value: V)
```
I don't think this needs to be a builder style method